_posts/2023-03-27-group-equivariant-cnn-2/group-equivariant-cnn-2.Rmd

```diff
@@ -0,0 +1,374 @@
+transformed_grid <- C_4$left_action_on_R2(C_4$inverse(g1)$unsqueeze(1), img_grid_R2)
+
+```
+
+Second, we re-sample the image on the transformed grid. The goat now looks up to the sky.
+
+```{r}
+transformed_img <- torch::nnf_grid_sample(
+  img$unsqueeze(1), transformed_grid,
+  align_corners = TRUE, mode = "bilinear", padding_mode = "zeros"
+)
+
+transformed_img[1,..]$permute(c(2, 3, 1)) |> as.array() |> as.raster() |> plot()
+```
+
+```{r, echo=FALSE, eval=TRUE, fig.alt = "Same goat, rotated up by 90 degrees."}
+knitr::include_graphics("images/z2.jpg")
+```
+
+## Step 2: The lifting convolution
+
+We want to make use of existing, efficient `torch` functionality as much as possible. Concretely, we want to use `nn_conv2d()`. What we need, though, is a convolution kernel that's equivariant not just to translation, but also, to the action of $C_4$. This can be achieved by having one kernel each for every possible rotation.
+
+Implement that idea is exactly what `LiftingConvolution` does. The principle is the same as before: First, the grid is rotated, and then, the kernel (weight matrix) is re-sampled to the transformed grid.
+
+Why, though, call this a *lifting convolution*? The usual convolution kernel operates on $\mathbb{R}^2$; our extended version operates on combinations of $\mathbb{R}^2$ and $C_4$. In math speak, it has been *lifted* to the [semi-direct product](https://en.wikipedia.org/wiki/Semidirect_product) $\mathbb{R}^2\rtimes C_4$.
+
+```{r}
+lifting_conv <- LiftingConvolution(
+    group = CyclicGroup(order = 4),
+    kernel_size = 5,
+    in_channels = 3,
+    out_channels = 8
+  )
+
+x <- torch::torch_randn(c(2, 3, 32, 32))
+y <- lifting_conv(x)
+y$shape
+```
+
+    [1]  2  8  4 28 28
+
+Since, internally, `LiftingConvolution` uses an additional dimension to realize the product of translations and rotations, the output is not four-, but five-dimensional.
+
+## Step 3: Group convolutions
+
+Now that we're in "group-extended space", we can chain a number of layers where both input and output are *group convolution* layers. For example:
+
+```{r}
+group_conv <- GroupConvolution(
+  group = CyclicGroup(order = 4),
+    kernel_size = 5,
+    in_channels = 8,
+    out_channels = 16
+)
+
+z <- group_conv(y)
+z$shape
+```
+
+    [1]  2 16  4 24 24
+
+All that remains to be done is package this up. That's what `gcnn::GroupEquivariantCNN()` does.
+
+## Step 4: Group-equivariant CNN
+
+We can call `GroupEquivariantCNN()` like so.
+
+```{r}
+cnn <- GroupEquivariantCNN(
+    group = CyclicGroup(order = 4),
+    kernel_size = 5,
+    in_channels = 1,
+    out_channels = 1,
+    num_hidden = 2, # number of group convolutions
+    hidden_channels = 16 # number of channels per group conv layer
+)
+
+img <- torch::torch_randn(c(4, 1, 32, 32))
+cnn(img)$shape
+```
+
+    [1] 4 1
+
+At casual glance, this `GroupEquivariantCNN` looks like any old CNN ... weren't it for the `group` argument.
```

Comment:
Again, here, if you're referring to the function, I'd add trailing parentheses.

Reply:
same thing here ... or "about same", actually ... (_about_, since I' talking about the argument even!)
So ... well if I want the first, I need this one for consistency ...